### ISSUE TITLE
CMake building system support

### ISSUE THREAD

**Issue by Wohlstand** [libxmp/libxmp#119]
Autotools event is a well system. However, is not so flexible for use and on some systems it may be hard to use. CMake is a flexible building system which is widely used by many various projects. It has lot of built-in features include various tweaks, finders, installers, downloaders, etc. It allows to generate regular Makefile for GNU-Make as well as project files for lots of various IDEs include Visual Studio, XCode, Code::Blocks and CodeLite., etc. CMake projects are can be easily opened in Qt Creator IDE which allows you to edit, test and debug your project easily as well as you will generate a project for any other IDE where you can debug stuff by using native dev-tools of given operating system.

I thinking to provide that by myself to build library as I will use it in my library project which widely uses CMake. I'll provide build at least for Linux, Windows and macOS operating systems, and also I'll try to build that for Emscripten thing and later for Android too.

**Comment by jaygarcia** [comment]
getting the lib to work with CMake was pretty simple. I'm no CMake Expert, but this worked well for me (within the context of a project using `SDL2`).

```
cmake_minimum_required(VERSION 3.10)

project(libxmp_project)

set(CMAKE_CXX_STANDARD 11)
set(CMAKE_FIND_LIBRARY_PREFIXES true)
add_compile_definitions(LIBXMP_CORE_PLAYER=true)
add_compile_definitions(LIBXMP_CORE_DISABLE_IT=true)


#set(EXPORTED_HEADERS gme/gme.h)

include_directories(
    ${SDL2_INCLUDE_DIR}
    ${CMAKE_CURRENT_SOURCE_DIR}
    "${CMAKE_HOME_DIRECTORY}"
    ${CMAKE_SOURCE_DIR}
    "${CMAKE_CURRENT_SOURCE_DIR}/libxmp/loaders/prowizard"
    "${CMAKE_CURRENT_SOURCE_DIR}/libxmp/loaders"
    "${CMAKE_CURRENT_SOURCE_DIR}/libxmp"
)


file(GLOB libxmp_SRC
    "libxmp/loaders/*.h"
    "libxmp/loaders/*.c"
    "libxmp/*.h"
    "libxmp/*.c"
)


add_executable(libxmp_project main.cpp  ${libxmp_SRC} libxmp/xmp.h ${libxmp_src})

target_link_libraries(libxmp_project ${SDL2_LIBRARIES})
```

**Comment by Wohlstand** [comment]
Oh, forgot this thread... Anyway, I'll also share my work I have tried to make:
https://github.com/WohlSoft/AudioCodecs/blob/master/libxmp/CMakeLists.txt

It's tested on Linux and wasn't tested on Windows and macOS yet, but I have kept a note for that.

P.S. the include that referred in this script is my generic manager of compiler flags I using to set optimization or shut up bogus warnings of some dirty compilers like MSVC: https://github.com/WohlSoft/AudioCodecs/blob/master/audio_codec_common.cmake

The purpose of CMake build I making is providing the same functionality that autotools does: platform and compiler features checks, and definition of necessary macros for the build.

**Comment by jaygarcia** [comment]
Thanks @Wohlstand =).

Your CMake example is more robust than mine.  What makes this too complicated for me (my example is a hacked version of the library) is the fact that there is a "lite" version of the Make implementation. I have no idea how to enable switching. 

I'm using the "lite" version to enable libXMP to work on the `ESP32` platform via `SPIRAM` for a video game we're developing.

Thank you for sharing your implementation =).

**Comment by Wohlstand** [comment]
Anyway, I'll extend mine with an option to choose what to build: lite or full. For now it's an attempt to build the full variant.

**Comment by Wohlstand** [comment]
Okay, I have polished my build script, and now it is buildable on Linux, Windows (even on MSVC, with a minor workaround in a form of the fake "unistd.h" header), on macOS and on Android.

**Comment by Wohlstand** [comment]
I had to polish a thing now https://raw.githubusercontent.com/WohlSoft/AudioCodecs/master/libxmp/CMakeLists.txt
- Now it can build both full and lite versions easily from one run, without copying the stuff between directories like the current Autotools build does
- I added some flags to tune the build:
  - `LIBXMP_DISABLE_IT` to disable IT support for Lite version
  - `LIBXMP_DISABLE_DEPACKERS` to disable depackers, I had to set it into ON by default as they won't build everywhere and fails to be built on several platforms
  - `LIBXMP_DISABLE_PROWIZARD` to disable the ProWizard support
  - `BUILD_STATIC` and `BUILD_SHARED` to choose which type of libraries to build
  - `LIBXMP_BUILD_LITE` and `LIBXMP_BUILD_FULL` to enable/disable building of lite or full versions

**Comment by AliceLR** [comment]
>   * `LIBXMP_DISABLE_DEPACKERS` to disable depackers, I had to set it into ON by default as they won't build everywhere and fails to be built on several platforms

Out of curiosity, which depackers fail to build for which platforms/toolchains? This sounds like a bug...

**Comment by sezero** [comment]
> > * `LIBXMP_DISABLE_DEPACKERS` to disable depackers, I had to set it into ON by default as they won't build everywhere and fails to be built on several platforms
> 
> Out of curiosity, which depackers fail to build for which platforms/toolchains? This sounds like a bug...

I don't how a build fail is possible unless it's a user fault.  We should at least see the failure messages.

**Comment by Wohlstand** [comment]
> Out of curiosity, which depackers fail to build for which platforms/toolchains? This sounds like a bug...

I don't remind, I should try to build them again on all my platforms.

> I don't how a build fail is possible unless it's a user fault. We should at least see the failure messages.

In my memory, the failure happens because of dependencies and headers missing, I'll retry to build them and check for actual errors.

**Comment by Wohlstand** [comment]
On Linux, I tested them immediately, builds just fine. Gonna test other platforms...

**Comment by Wohlstand** [comment]
On macOS it also works, on Windows, my build at AudioCodes does work with MinGW and MSVC. Just now I reminded that on Windows MSVC they all will fail because of the `unistd.h` missing.

**Comment by Wohlstand** [comment]
It gets fixed once I added the minimal custom `unistd.h` at me for MSVC as a workaround: `https://github.com/WohlSoft/AudioCodecs/blob/master/libxmp/include_msvc/unistd.h`

**Comment by sezero** [comment]
> It gets fixed once I added the minimal custom `unistd.h` at me for MSVC as a workaround: https://github.com/WohlSoft/AudioCodecs/blob/master/libxmp/include_msvc/unistd.h

As I said, it's user error: a dummy unistd.h already exists in mainstream for msvc builds.

**Comment by Wohlstand** [comment]
Okay, I see:
https://github.com/libxmp/libxmp/blob/882779777f74d32c8e7b5d901a985c033072855a/src/win32/unistd.h
Gonna tune my build.

**Comment by Wohlstand** [comment]
Hi there!

Suddenly, I reminded about this issue, and I think, I need to backport my CMake code into the mainstream. I had to test it on various platforms and hardware things (people had to build my code on various operating systems and having different hardware platforms include various types of ARM and the PPC64LE architecture). I want to get feedback on my CMake code if there is something I need to add, etc. Right now my code is intended to work as a part of the AudioCodecs repository where most libraries to refer the global CMake file that includes the set of compiler optimizers and helper variables, so, I'll tune that too to make the code being friendly with the mainstream.

Also, Visual Studio since 2017 is able to open CMake projects directly.
